### ISSUE TITLE
Suggestion: Provide a simple example of actually async-parsing a CSV file

### ISSUE THREAD

**Issue by kennu** [adaltas/node-csv#267]
It is currently difficult for a first-time user to understand how to use the CSV module because there isn't a simple example of how to actually parse a file. I think most users are not interested in the generator functionality but just parsing a file using modern async code, so I would leave the generator stuff out.

I would suggest adding an example to the "Getting Started" section: If you have a file called "file.csv", here's how you use the async parser to parse it.

My understanding of how to do it is like this, I'm happy to hear if it's not the right way:

```typescript
import * as csvParse from 'csv-parse'
import { createReadStream } from 'fs'

async function main()  {
  const parser = csvParse()
  createReadStream('file.csv').pipe(parser)
  for await (const row of parser) {
    console.log(row)
  }
}

main()
.then(() => {})
.catch(console.error)
```

**Comment by wdavidw** [comment]
Using the async iterator API is documented [here](https://csv.js.org/parse/api/async_iterator/). Note, you can simplify your code with `const parser = createReadStream('file.csv').pipe(csvParse())`.

This library is not about using `createReadStream`, nor about using the `fs` API, even popular and useful. This is why we introduced [recipes](https://csv.js.org/parse/recipies/). Feel free to add/modify/enrich a new one. With more recipies, we will provide more examples adjusted to everyone use cases.

Also, there are many ways to use the library, not just the one you are suggesting (which is a very good one). This is the reason why the Getting Started page must remain open. Users interested parsing CSV records can refer to the Parse documentation and disregard the Generator one. I personnaly use it a lot to simulate Streaming and IoT workloads.

**Comment by kennu** [comment]
Thanks for your response. I read exactly the page you linked, and didn't see a simple example of how to parse a CSV file, so that is why I made my suggestion. My intention was simply to make it easy for a new user to understand how to parse a CSV file without having to dive into deeper documentation.

**Comment by MarkPare** [comment]
@kennu I ran into the same issue. @wdavidw I see your point, but in practice I think people would like to see this in an easily-discoverable part of the docs (e.g. in [Examples](https://csv.js.org/project/examples/) section) since it's such a common use case.

Anyway, below code works for me:

```
import parse from 'csv-parse';
import fs from 'fs';
 
 const processFile = async () => { 
  return new Promise((resolve, reject) => {
    const path = 'file.csv';
    const stream = fs.createReadStream(path);
    const parser = parse();

    stream.on('ready', () => {
      stream.pipe(parser);
    });

    parser.on('readable', function(){
      let record;
      while (record = parser.read()) {
        // Do transforms / side effects here
        console.log('record', record);
      }
    });

    parser.on('error', function(err){
      console.error(err.message);
      reject();
    });

    parser.on('end', function(){
      console.log('Parsing complete');
      resolve();
    });
  });
 }

 processFile();
```

**Comment by wdavidw** [comment]
I have in the pipe the idea to create a "./lib/promise.js" module every package of csv, but it has been a few years now and never found the time.

**Comment by wdavidw** [comment]
I think your example could be simplified:

```js
import parse from 'csv-parse';
import fs from 'fs';
import stream from 'stream/promises';
 
 const processFile = async () => { 
  const stream = fs.createReadStream(path);
  const parser = parse();
  parser.on('readable', function(){
    let record;
    while (record = parser.read()) {
      // Do transforms / side effects here
      console.log('record', record);
    }
  });
  await finished(rs);
 }

 processFile();
```

Do you confirm ? That would be a nice additionnal recipe.

**Comment by MarkPare** [comment]
Yeah looks good. I was having issues a couple hours ago with the stream and using the `ready` event listener helped, but I just confirmed it's not necessary.  Only necessary tweak I see is making sure `path` is defined.

**Comment by wdavidw** [comment]
By the way, I never use those but async iterator is even more elegant:

```js
const processFile = async () => {
  records = []
  const parser = fs
  .createReadStream(`${__dirname}/fs_read.csv`)
  .pipe(parse({
    // CSV options if any
  }));
  for await (const record of parser) {
    // Work with each record
    records.push(record)
  }
  return records
}
```

**Comment by buuug7** [comment]
the package's document is so confuse, all example is about generator and then parse, why not have a example with read from file.csv? the api is very confuse me, not clear, most people see the get started example is very confuse and give it up.

**Comment by wdavidw** [comment]
It all depends on wether you are looking for simplicity or scalability. Did you look at the [sync API](https://csv.js.org/parse/api/sync/), it is the easiest way to use the parser. If you are looking for scalability, meaning loading data which is not meant to be entirely in memory, you need to use either the Node.js stream API or the generator API as above. Using async/await is more complicated than the Sync API and is not scalable like the stream and generator API. Thus, we do not provide an API for it.

**Comment by buuug7** [comment]
@wdavidw you should see the react document, it have a big code base, but api is simplified as blow:

```javascript
class HelloMessage extends React.Component {
  render() {
    return (
      <div>
        Hello {this.props.name}
      </div>
    );
  }
}

ReactDOM.render(
  <HelloMessage name="Taylor" />,
  document.getElementById('hello-example')
);

you should think again.
```

**Comment by buuug7** [comment]
@wdavidw you should give the user default options other than with verbose options for each api.
I think it should be **An opinionated CSV parser**.

**Comment by buuug7** [comment]
@wdavidw the more you need , the less you have. plese keep it simple and clear. the document one moment with generator, one moment stream, one momment callback , one momment async.

**Comment by wdavidw** [comment]
We deal with large CSV files daily and multiple of them in parallel, most of the time we can't load them entirely in memory. The library is 10y old, don't think its API comes from the sky. The default API works and will always work no matter the input size and the use case (eg batch versus steaming). It is not the easiest but it works. Not the case of the sync API, that is however easier, or an eventual async/await, that does not make sense.

**Comment by buuug7** [comment]
@wdavidw I'm a little over the top. I'm sure your work is excellent. Sorry, it's never easy to write code.

**Comment by buuug7** [comment]
@wdavidw but the document need simplied and cover use case, the api need keep up same style.

**Comment by wdavidw** [comment]
I totally agree, there is a lot of room to improve the documentation, particularly in that regard. Either using the same API, or clarify the one we used with a link.

**Comment by buuug7** [comment]
@wdavidw compared with react document and prettier document , the node-csv is not good, but node-csv documentation is better than most other libraries .

**Comment by julien-monchany** [comment]
> By the way, I never use those but async iterator is even more elegant:
> 
> ```js
> const processFile = async () => {
>   records = []
>   const parser = fs
>   .createReadStream(`${__dirname}/fs_read.csv`)
>   .pipe(parse({
>     // CSV options if any
>   }));
>   for await (const record of parser) {
>     // Work with each record
>     records.push(record)
>   }
>   return records
> }
> ```

This example should definitely be in the API docs

**Comment by wdavidw** [comment]
There is one: https://csv.js.org/parse/api/async_iterator/

**Comment by t-fritsch** [comment]
thanks @wdavidw for the great work on this lib, it is apparently the most maintained and mature csv lib that exists, congrats for that !

I came here because I couldn't find a "simple enough for me" example of how to use this lib to parse a file in async/await function.

This issue helped me a lot but the example you provided in your previous message (_"There is one: https://csv.js.org/parse/api/async_iterator/"_) isn't as good as the one you posted here https://github.com/adaltas/node-csv/issues/267#issuecomment-730256217 (no generator + reads file from disk). \
Anyway I found out that this example does exist in the doc, but in the recipes section : https://csv.js.org/parse/recipes/async_iterator/

I think this recipes section is **really usefull** but maybe discovered too late by newcomers, and maybe this could be solved by some ways : 
- move recipes up in the left menu (for now it is the last section, not visible if not scrolling through every options)
- having a minimal example in the "Usage" page + the full recipes list right in this page
- mention the related recipes in the corresponding API pages (for example mention https://csv.js.org/parse/recipes/async_iterator/ in https://csv.js.org/parse/api/async_iterator/) or even better : include recipes right in the API page, as examples !

**Comment by wdavidw** [comment]
Thank you @t-fritsch for your feedback, here are my comments:

> move recipes up in the left menu 

OK

> having a minimal example in the "Usage" page + the full recipes list right in this page

I'll rather just list the recipes, there are too many divergent expectations for using the lib. However, I dislike having to maintain the list up-to-date in a section of the website that I will forget the next day.

> mention the related recipes in the corresponding API pages

OK

> include recipes right in the API page, as examples

APIs are about how to call the lib depending on your usage, recipes more about it usage and integration. However, I just realized there are 2 pages covering the async iterator. It is not on purpose.

**Comment by pbrink231** [comment]
I saw the example @wdavidw linked to and it looks nice but does now work in my case I believe.

I am using basic-ftp that takes a write stream.  I am pulling a large file and would like to process it in chunks so it does not overload memory.

I have this as my setup
```
  passThroughPipe
    .pipe(csvParser({ separator: '|' }))
    .on('data', async (data) => {
      console.log('got data', data);
      processedData.push(data);
      await new Promise((r) => setTimeout(r, 2000));
      console.log('done waiting');
    })
    .on('end', () => {
      console.log('done');
      console.log('processedData', processedData);
    });

    const processReturn = await client.downloadTo(
      passThroughPipe,
      remoteFilepath
    );
```

Is there not a simple way to allow .on('data') be async.  Maybe include a next call back on it that stops the flow until it finishes?  Sorry if there is an obvious way to rewrite this, am still new to streams.

My timout is just to simulate some async operation that needs to happen.

**Comment by pbrink231** [comment]
Follow this up with a solution I think would be good for the docs.

```
  const testParser = csvParser({ separator: '|' });
  testParser.on('data', async (data) => {
    // pause reading and backpressure
    testParser.pause();

    // Some async operation
    await processor(data);

    // make sure to resume operations
    testParser.resume();
  });
  testParser.on('end', () => {
    console.log('done');
  });

    const processReturn = await client.downloadTo(testParser, remoteFilepath);
    console.log('processReturn', processReturn);
```

**Comment by wdavidw** [comment]
For some reason, I received 3 notifications from GitHub with comments associated to this issue which I cannot browse once I open the page. @pbrink231 if you see this message, please open a new issue.

**Comment by kenn** [comment]
Here's my version — simplest of all?

```js
async function main() {
  createReadStream(`${__dirname}/fs_read.csv`).pipe(
    csv.parse(async (err, data) => {
      for (const row of data) {
        await doAwaitableTask(row)
      }
    }),
  )
}
main()

```

**Comment by wdavidw** [comment]
There is now an example illustrating fs operations here: https://csv.js.org/project/examples/file_interaction/
